pkg/controller/certificates/inconflict/inconflict_controller.go

```diff
@@ -0,0 +1,276 @@
+	mustSync := []cache.InformerSynced{certificateInformer.Informer().HasSynced}
+
+	return &controller{
+		certificateLister: certificateInformer.Lister(),
+		client:            ctx.CMClient,
+		fieldManager:      ctx.FieldManager,
+		queue:             queue,
+	}, queue, mustSync
+}
+
+// ProcessItem is the reconcile loop which manages the InConflict
+// condition on Certificate resources. This condition signals that the
+// Certificate shares the same spec.secretName as another Certificate in the
+// same Namespace.
+func (c *controller) ProcessItem(ctx context.Context, key string) error {
+	log := logf.FromContext(ctx).WithValues("key", key)
+
+	ctx = logf.NewContext(ctx, log)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.Error(err, "invalid resource key passed to ProcessItem")
+		return nil
+	}
+
+	crt, err := c.certificateLister.Certificates(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		// If the Certificate is not found, we can assume it has been deleted
+		// unfortunately. We cannot rely on the resource's InConflict
+		// condition to determine what Certificates it was conflicting with, so
+		// we need to loop over all Certificates in the Namespace and requeue
+		// Certificates which have this Certificate's name in their
+		// InConflict condition.
+
+		// Get all Certificates in the Namespace.
+		crts, err := c.certificateLister.Certificates(namespace).List(labels.NewSelector())
+		if err != nil {
+			return err
+		}
+
+		// Loop over all Certificates and requeue those which have this
+		// Certificate's name in their InConflict condition.
+		for _, crt := range crts {
+			if condition := apiutil.GetCertificateCondition(crt, cmapi.CertificateConditionInConflict); condition != nil &&
+				condition.Status == cmmeta.ConditionTrue &&
+				stringSliceContains(strings.Split(condition.Reason, ","), name) {
+				c.queue.Add(crt.Namespace + "/" + crt.Name)
+			}
+		}
+
+		return nil
+	}
+	if err != nil {
+		log.V(logf.DebugLevel).Info("certificate not found for key", "error", err.Error())
+		return nil
+	}
+
+	log = logf.WithResource(log, crt)
+	ctx = logf.NewContext(ctx, log)
+
+	oldDuplicates := []string{}
+	if condition := apiutil.GetCertificateCondition(crt, cmapi.CertificateConditionInConflict); condition != nil &&
+		condition.Status == cmmeta.ConditionTrue {
+		oldDuplicates = strings.Split(condition.Reason, ",")
+	}
+
+	// Get the Certificates in the same Namespace which have the same Secret name
+	// set.
+	duplicates, err := internalcertificates.DuplicateCertificateSecretNames(ctx, c.certificateLister, crt)
+	if err != nil {
+		return err
+	}
+
+	// If we don't need to update the status, return early.
+	if !needsUpdate(crt, duplicates) {
+		return nil
+	}
+
+	condition := buildInConflictCondition(duplicates)
+	crt = crt.DeepCopy()
+	if condition != nil {
+		apiutil.SetCertificateCondition(crt, crt.Generation, condition.Type, condition.Status, condition.Reason, condition.Message)
```

Comment:
**Note:** The `reason` field is meant to be a one-word CamelCase reason for the condition's last transition, as per https://github.com/kubernetes/community/blob/4c9ef2d/contributors/devel/sig-architecture/api-conventions.md

Reply:
Thanks, currently, the reason is used to store what certs are in conflict & this value is actually used by the controller.
As discussed in today's standup, we probably can implement a solution that does not require this.